### ISSUE TITLE
Comments on the architecture.md

### ISSUE THREAD

**Issue by iskrisis** [hackersanddesigners/cosmogrammatics#1]
Kirby officially doesn't have block IDs yet. It's feature that should come probably in 3.9. so the ids you see are internal ids and the mechanisms will probably change. I think they are ephemeral so that's dangerous to use.

This is going to be honestly a bit pain because even the pre 3.8. https://github.com/bnomei/kirby3-autoid didn't work with blocks. You could probably persist the block id kirby currently creates to the fields on hook when saving the page and then use it. But it not something i've tried.

I also think you would probably need the Sqlite even with like hundreds comments. Using virtual pages it can be done so the comments keep the kirby API. The main reason is relations. Even with the UUID system from 3.8. it helps performance of getting individual page/comment when you know the id (so like PARENT) but it's still bad when you want the opposite relation (getting all comments that have ID as parent). You would probably end up keeping the list of childrenIDs on parent too but then you have to always handle this when you create/update anything. A bit tedious. It's nosql problems 🤷

**Comment by sonn-gamm** [comment]
Thanks for sharing your insight, wonder how you came across this repo? (: 

I was not aware that block's id are not "stable" yet. I shared this on the forum and no one said anything about block ids being not finalized (https://forum.getkirby.com/t/kirby-blocks-comments-saved-in-sqlite/26256).

From you suggestion: you mean I should use the Panel after-update hook to permanently save the internal block id to an explicit field? How do I keep the relation between the field and the original block id?

Right now when creating an article in reference to a specific block, we save the block id in a comment field. Is this what you mean?

I'll have another look at integrating SQLite with Kirby. I got a bit put off at having to learn more PHP than necessary in order to setup SQLite decently (eg doing migration, initial table creation, etc), and I moved away from the language in the last few years and would like to not necessarily touch it anymore — though Kirby is a great piece of software. But I can maybe do all the SQLite maintenance stuff from the command line.

**Comment by iskrisis** [comment]
Don't forget gh is also a social network. I've been tracking what you and h&d are doing for quite some time.
I've briefly met Anja when she had presentation in Bratislava (i am czech - close) and i saw Magdalena last week and she mentioned you so i got reminded. I just saw you pop up in my timeline on gh i checked what are you up to nowdays and to my surprise i see Kirby.

I meant what you are probably doing. Save the id to some hidden field on the block object in a hook. To be sure you persist it since it may change. You could source dive current UUID algo to see how its created (with kirby being so straightforward its usually very easy) to see if they may change.

The UUID situation is a bit complicated. Blocks came in K3.6 and i think internally needed UUIDs for their functionality. But UUIDs were also the most requested feature - it has rolled out in 3.8 (which is like two weeks old) but only for pages, users and files. Not blocks or structures. That should come later and the structure of those are `page://hb38HvnQfm8HlQ6e` `files://hb38HvnQfm8HlQ6e` so i assume they will change the block UUIDs to `block://hb38HvnQfm8HlQ6e` in like 3.9.

Conceptually the way virtual pages work (kirby calls every data grouping page) is that you overwrite the methods on the page object that read/write to the filesystem. It's actually quite elegant (albeit not that documented), its just few methods that you need to overwrite. And you can do anything with the reading/saving data - transform it, save it to multiple places... kind of middleware situation. Good trick here is to make columns for the important/indexd stuff titles, ids, relations and then take everything else that you don't have column for and slap it into one sqlite JSON column. This way it will kinda work as kirby without needing to go back to change the db schema/migrate too much.

I wouldn't worry about sqlite. Even in the default mode it can handle writes very quickly. But good idea is to put it into WAL mode. It will have speed and scale that i don't think you can possibly hit. Kirby DB drivers are bare bones, it abstracts the connection setup and allows you to run queries thats about it. So no ORMs. So i don't think you would be writing too much PHP more like a lot of raw SQL. The migrations suck but it will probably not happen that often and like you said you can do that in some sqlite gui tbh (thats how i do it). You can possibly start with the filesystem but some of the relation code would be different.

I understand you on PHP. I am probably a big brute in this and all the popular interpreted dynamic languages PHP/JS/Python/Ruby kinda seem same to me. But i understand the aesthetics are terrible. Problem is I have yet to find anything that is so straightforward, productive and also safe to host. I have many tiny websites/services on single server that can't affect each other. And Kirby is just really so focused and purposeful. Please if you know about something so utilitarian let me know (i try everything i can get my hands on). For example i would really love to learn some functional programming systems like elixir/erlang. I just don't think its worth it to me just as hobby as the don't even have CMSes. JS cmses are trying to catch up but its still mostly too abstracted complexity.

Kirby has pretty active discord where you can ask about more advanced things other devs. Surely check it out if you don't mind chats.

**Comment by sonn-gamm** [comment]
I am aware of the core GH motivations, all good. As I stopped being active much in here, I was pleasantly surprised that someone took the time to look at the repo and share suggestions!

You almost convinced me to give the SQLite option another try, will take a look in the coming days. Thanks!

Kirby is super documented but sometimes also not properly or enough for some things. See this block id thing. I also stopped hanging out in the Kirby forum much unlike years ago, so have no idea what's in their roadmap.

Re: Languages. Agree.
